File: src/api/restaurants.ts
import { defaultFetcher, type Fetcher } from "@literate.ink/utilities";
import { Request } from "~/core/request";
import { decodeRestaurant } from "~/decoders/restaurant";
import { type Meal, Moment, type Restaurant } from "~/models";

export async function restaurants (identifier: string, fetcher: Fetcher = defaultFetcher): Promise<Array<Restaurant>> {
  const request = new Request(`${identifier}/externe/crous-${identifier}.min.json`);
  const response = await fetcher(request);

  const content = response.content.replace(/[\u0000-\u001F]/g, "");
  const { restaurants } = JSON.parse(content);

  return restaurants.map(decodeRestaurant);
}

export function isRestaurantOpen (restaurant: Restaurant, dayIndex: number, moment: Moment): boolean {
  const day = restaurant.opening.split(",")[dayIndex];
  const momentAsINT = moment === Moment.LUNCH ? 1 : moment === Moment.MORNING ? 0 : 2;
  const opening = day[momentAsINT];

  return opening === "1";
}

export function meals (restaurant: Restaurant, date = new Date()): Array<Meal> | undefined {
  const currentSTR = date.toLocaleDateString();
  return restaurant.menus.find((menu) => menu.date.toLocaleDateString() === currentSTR)?.meals;
}

File: library/src/commonMain/kotlin/ink/literate/crowous/api/restaurants.kt
package ink.literate.crowous.api

import ink.literate.crowous.core.Request
import ink.literate.crowous.decoders.decodeRestaurant
import kotlinx.datetime.LocalDate
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import ink.literate.crowous.models.Meal
import ink.literate.crowous.models.Moment
import ink.literate.crowous.models.Restaurant

suspend fun restaurants (identifier: String): List<Restaurant> {
  val request = Request("$identifier/externe/crous-$identifier.min.json")
  val response = request.send()

  val content = response.replace(Regex("[\u0000-\u001F]"), "")
  val json = Json.parseToJsonElement(content)

  return json.jsonObject["restaurants"]!!.jsonArray.map { decodeRestaurant(it.jsonObject) }
}

/**
 * Check if a restaurant is open at a specific day and moment.
 *
 * @param restaurant The restaurant to check.
 * @param dayIndex The index of the day to check, starting from 0 for Monday.
 * @param moment The moment to check.
 */
fun isRestaurantOpen (restaurant: Restaurant, dayIndex: Int, moment: Moment): Boolean {
  val day = restaurant.opening.split(",")[dayIndex]
  val opening = day[moment.ordinal]

  return opening == '1'
}

/**
 * Get the meals of a restaurant for a specific date.
 *
 * @param restaurant The restaurant to get the meals from.
 * @param date The date to get the meals for.
 */
fun meals (restaurant: Restaurant, date: LocalDate): List<Meal>? {
  return restaurant.menus.find {
    it.date.monthNumber == date.monthNumber
        && it.date.dayOfMonth == date.dayOfMonth
  }?.meals
}
